web/modules/custom/pulses_exchange_rate/src/Plugin/Block/PulsesExchangeRate.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace Drupal\pulses_exchange_rate\Plugin\Block;
+
+use Drupal\Core\Block\Annotation\Block;
+use Drupal\Core\Block\BlockBase;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\pulses_exchange_rate\CurrencyDataProviderInterface;
+use Drupal\pulses_exchange_rate\Form\ExchangeApiKey;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a 'Exchange Rates' block.
+ *
+ * @Block(
+ *   id = "exchange_rates",
+ *   admin_label = @Translation("Exchange Rates"),
+ * )
+ */
+class PulsesExchangeRate extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * Instance of CurrencyDataProvider.
+   *
+   * @var \Drupal\pulses_exchange_rate\CurrencyDataProvider
+   */
+  public $client;
+
+  /**
+   * Instance of ConfigFactoryInterface.
+   *
+   * @var \Drupal\pulses_exchange_rate\CurrencyDataProvider
+   */
+  public $configFactory;
+
+  /**
+   * Construct PulsesExchangeRate class.
+   *
+   * Get instance of Client class.
+   *
+   * @param array $configuration
+   *   Current plugin configuration.
+   * @param string $plugin_id
+   *   Current plugin id.
+   * @param mixed $plugin_definition
+   *   Current plugin definition.
+   * @param \Drupal\pulses_exchange_rate\CurrencyDataProviderInterface $currencyProvider
+   *   Instance of currencyProvider class.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, CurrencyDataProviderInterface $currencyProvider) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->client = $currencyProvider;
+    $this->configFactory = $currencyProvider->configFactory;
+  }
+
+  /**
+   * Injecting dependencies.
+   *
+   * @param \Symfony\Component\DependencyInjection\ContainerInterface $container
+   *   {@inheritDoc}.
+   * @param array $configuration
+   *   Current plugin configuration.
+   * @param string $plugin_id
+   *   Current plugin id.
+   * @param mixed $plugin_definition
+   *   Current plugin definition.
+   *
+   * @return static
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('pulses_exchange_rate_service'),
+      $container->get('config.factory'),
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build() {
+    $api_key = $this->configFactory->getEditable(ExchangeApiKey::SETTINGS)->get('key');
+    if (!empty($api_key)) {
+      $this->client->getResponse('https://api.currencyapi.com/v3/latest?apikey=', $api_key, TRUE);
```

Comment:
An interesting question to think about:
What happens when you have 100 thousand visitors to your site?

Reply:
Cache for Api data was added.